### DISCUSSION TITLE
offset calculation incorrect?

### DISCUSSION THREAD

**Discussion by sleveille4** [jmcollin78/versatile_thermostat#236] [General]
Hi,
i set up a versatile thermostat to operate a TRV and "correct" it's internal temp sensor, replacing it by another more accurate sensor within the room

but it seems that the calculation of the offset correction is incorrect and as a result i cannot reach the desired T° in the room.
here is an example:
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/1c168069-6141-4c24-89c9-26245951be59)

on the left, is the TRV status, in center the versatile to "correct" the TRV, and on the right the 2 relevant T° sensors

as you can see, the TRV reports 18.8°C temp, the room sensor 15.9°C, and i set versatile thermostat to 17°C, it then only send 18.5°C target temp to the TRV.

thus the TRV do not open (because it's target is below it's sensor), althought the sensor temp in versatile is 1°C BELOW versatile target temp

the versatile entity is configured as follow:

![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/e5a30670-c70a-484a-a6c3-c82ee9609ee0) 
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/085e5208-ceb4-4e20-8f5a-e72dca53bd0d)
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/044819d0-02e9-47c1-9895-f55bee8bb69c)
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/87e16a0b-3eb4-4315-b4d1-499b5adc9526)

i indeed waited several cycles but it always seems to work "inverted", the target temp sent to the TRV is always below the TRV sensor

**Comment by maia** [comment]
I'm unsure about the exact offset limits, so it's possible that the regulated target temperature is restricted to +1,5°C over the target temperature. But what you're experiencing is that versatile_thermostat does not know or care about the temperature measured inside the TRV. No one has yet come up with an algorithm that deals with the room temperature sensor *and* the TRV internal sensor.

**Reply by sleveille4** [reply]
taht's odd given that the main point is to actually replace the inacurate (by design because it is so close to the heat source) sensor of the TRV with an outside sensor, by checking the difference between them and applying the proper offset. thus it requires to actually read the TRV's sensor

**Reply by jmcollin78** [reply]
Hello, can you please try the Strong regulation mode. It should work better for your case.

**Comment by sleveille4** [comment]
hi,
switched regulation mode to "strong", will see how it goes in a few hours, but at first glance, the issue seems to remain:
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/7f177495-c18e-4ff4-b744-e2c3a1059e80)

**Comment by jmcollin78** [comment]
This is strange.

If I understand well, the target t° is 17,5 and the current temp is 16,6°. But the internal TRV current temp is 19° (+ 2,4 ° above the real temp if I understand which is huge gap). In strong regulation, the offset can go to up to 5°, so it should definitively heat. 

With the https://github.com/jmcollin78/versatile-thermostat-ui-card you will be able to control the regulated temperature and the regulation mode. And in the attributes of the VTherm you will have some useful information. Can you please post them here during the issue ? (they are available in Dev Tools/States. Copy paste the yaml of the attributes formatted with the <> button).

The other thing I will need to try to understand is to have a curve with target t°, regulated t°, current t°. See an example here: https://github.com/jmcollin78/versatile_thermostat/tree/4.2.0#even-better-with-apex-chart-to-tune-your-thermostat

**Reply by sleveille4** [reply]
hi, will try that later & post results, as of now the outside T° went up again thus i turned the boiler off and this thermostat is no longer on either.

as for the delta between the TRV T °sensor and the other T° sensor, it is by default no that high, everything is off since yesterday night and the difference between the 2 is around 1°C. 
however the difference drastically increase when the radiator is hot as the TRV is directly ON the radiator, thus receives a lot of heat.
this is actually the reason i want to use an external T° sensor, so that the TRV stop being disturbed by the heat generated by the radiator and uses proper room temperature instead

**Comment by sleveille4** [comment]
as requested the graph (6h span)
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/695ec412-c6cf-42b2-a773-7f772c04a66a)

i installed the UI card and here is a cap of the card (on the right is the TRV card) at the moment the graph ends, where the issue is again happening
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/ab8bc8d3-a034-403d-856b-7def70120733)

and the attributes of the versatile thermostat:
`hvac_modes: off, heat
min_temp: 4
max_temp: 35
preset_modes: none, eco, comfort, boost
current_temperature: 14.4
temperature: 15
hvac_action: heating
preset_mode: none
is_on: true
hvac_mode: heat
type: null
eco_temp: 14
boost_temp: 17.5
comfort_temp: 16
eco_away_temp: null
boost_away_temp: null
comfort_away_temp: null
power_temp: null
target_temperature_step: null
ext_current_temperature: 5.6
ac_mode: false
current_power: null
current_power_max: null
saved_preset_mode: eco
saved_target_temp: 15
saved_hvac_mode: null
window_state: off
motion_state: null
overpowering_state: null
presence_state: null
window_auto_state: off
window_bypass_state: false
security_delay_min: 60
security_min_on_percent: 0.5
security_default_on_percent: 0.1
last_temperature_datetime: 2023-12-07T10:30:34.168155+01:00
last_ext_temperature_datetime: 2023-12-07T10:01:11.036671+01:00
security_state: false
minimal_activation_delay_sec: 10
device_power: 1
mean_cycle_power: null
total_energy: 9.39
last_update_datetime: 2023-12-07T10:43:35.403336+01:00
timezone: Europe/Paris
window_sensor_entity_id: null
window_delay_sec: null
window_auto_enabled: false
window_auto_open_threshold: null
window_auto_close_threshold: null
window_auto_max_duration: null
motion_sensor_entity_id: null
presence_sensor_entity_id: null
power_sensor_entity_id: null
max_power_sensor_entity_id: null
temperature_unit: °C
is_device_active: true
ema_temp: 14.54
is_over_climate: true
start_hvac_action_date: null
underlying_climate_0: climate.thermostat_sdb_thermostat
underlying_climate_1: null
underlying_climate_2: null
underlying_climate_3: null
is_regulated: true
regulated_target_temperature: 16.2
auto_regulation_mode: auto_regulation_strong
regulation_accumulated_error: 13.00000000000001
friendly_name: versatile_sdb
supported_features: 17`


also, unrelated, but it seems that the versatile thermostat is "stuck" in hvac_action"heating" instead of adopting the TRV status since yesterday's update (had to switch the source of "heating" status from Vtherm to the actual TRV in the graph to have accurate heating periods)

**Comment by sleveille4** [comment]
quick update as the vtherm ran for a few more minutes, it eventually triggered heating (zoomed in the graph to improve visibility), 
![image](https://github.com/jmcollin78/versatile_thermostat/assets/25745455/da9e524b-b9df-4de8-9bb4-af3016fec02d)



but as you can see from the graph, it only triggered heating when T° reached 14.3°C, then heated for +-20m, and stopped heating when temp reached 14.7°C.
the target being 15°C, regulation threshold is set to 0.3°C, and cycle being 5m (for test purpose, i intend to bump it to 10 or 15m later, to save TRV battery as explained in another thread)

with such parameters, i would have expected the heating to start around 10:00 to 10:10 (at least one cycle started within this timeframe)
also once heating started, it seems that it stopped short of the target by the threshold amount, i woud have expected it either to reach target, or overshoot target by thereshold (I.E stop heating when 15°C is reached, or 15.3°C is reached)

**Reply by jmcollin78** [reply]
In this last screen, what I see in normal:
- at 10:50:
VTherm target is 15°, but room temp is 14,3° so the regulated temp send to the TRV is a few higher 16,5° (you have accumulated some errors because before 10:50 room temp is < target temp. So there is +1.5 ° regulation from VTherm.
the TRV have for target 16,5° (the VTherm regulated temp) so it is normal but because the internal temp of the TRV is 15.6°, it is heating.

- at 11:10 the TRV internal temp is >> target temp so it stops heating and the real room temp is 14.5 so near the target.

What I see here is that the real room target will oscillate around the target and that regulation does its work by forcing heating.

All seems normal.

I guess you have a cycle of 30 min in this last test no (the delta time between 2 regulation changes) ? If yes, you could certainly lower the cycle_min to 15 min to have more precision and reactivity.

**Reply by jmcollin78** [reply]
The only think not normal is the gap in the TRV current temp just before it starts heating. 
![Capture d’écran 2023-12-08 à 16 54 44](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/54ff620b-ef23-438c-a0cb-83868130b6cc)

**Reply by jmcollin78** [reply]
We can certainly try more regulation in your case to start heating earlier.
But you have to use the expert mode for this. You can try this in expert mode:

```
versatile_thermostat:
  auto_regulation_expert:
    kp: 0.5
    ki: 0.1
    k_ext: 0.02
    offset_max: 6
    stabilization_threshold: 0.1
    accumulated_error_threshold: 50
``` 

This configuration is to be put into the `configuration.yaml` file and set regulation to 'Expert'

**Comment by sleveille4** [comment]
what i do find strange in my graph is that the heating doesn't start around 10:00 to 10:10. the room temp is clearly below the target, so i would have expected vtherm to recompute a delta and apply it as to start heating, but it does not.
it waits until 10:25, and still it fall "short" as TRV target is increased, but still remains below TRV temp, thus it doesn't heat either.

the sudden drop you mention at 10:45 is also a mystery to me, the only explanation i can find is that TRV stopped reporting temp for some time then suddenly updated.

TRV temp stay flat @ 16.1°C for around 30 minutes while during this time the room sensor keep decreasing at +- a steady pace

as for the time, i still haven't changed it yet, it is still on a 5 minute basis for test purpose.

will try the expert setting, to be clear i just have to enter the setting once and it will be applied to all vtherm set to expert correct?

i'm asking because i received the other 6 TRV and i intend to install them this weekend, so far i just had 1 for test purpose.

**Reply by jmcollin78** [reply]
> what i do find strange in my graph is that the heating doesn't start around 10:00 to 10:10.  the room temp is clearly below the target

I see the exact inverse. Current is 16°5 and target is 16° :
![Capture d’écran 2023-12-09 à 07 50 03](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/b78484ef-63a6-46e1-9d97-903e32e565ca)

>  i just have to enter the setting once and it will be applied to all vtherm set to expert correct?

Yes. You add the code into `configuration.yaml` restart HA (or just the VersatileThermostat integration) and then change the VTherm regulation mode to expert. I'm confident into this settings to solve your point.
